### ISSUE TITLE
twisted.test.test_udp.MulticastTestCase.test_multiListen implicitly expects multicast group participation to be host-wide not per-socket

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5090]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5090|
|Type|enhancement|
|Created|2011-05-06 21:30:29Z|
|Branch|https://github.com/twisted/twisted/tree/multilisten-5090|

The test, [shown here](http://twistedmatrix.com/trac/browser/trunk/twisted/test/test_udp.py?rev=30752#L659), has the server join a group but neither client join a group.  Then it expects clients to receive messages from that group.  This is, of course, silly, and it's a real surprise that it works most of the time.  As a bonus, the test's failure mode is to hang forever if it doesn't work.

<details><summary>Searchable metadata</summary>

```
trac-id__5090 5090
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_multilisten_5090 branches/multilisten-5090
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1304717429000000 1304717429000000
changetime__1318726488000000 1318726488000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Here's a patch.  I guess this might even be good enough for review?  Assigning it to myself to think about a little more before I commit to that though.

```
#!diff
=== modified file 'twisted/test/test_udp.py'
--- twisted/test/test_udp.py	2011-02-14 04:45:15 +0000
+++ twisted/test/test_udp.py	2011-05-06 21:48:26 +0000
@@ -6,7 +6,7 @@
 Tests for implementations of L{IReactorUDP} and L{IReactorMulticast}.
 """
 
-from twisted.trial import unittest, util
+from twisted.trial import unittest
 
 from twisted.internet.defer import Deferred, gatherResults, maybeDeferred
 from twisted.internet import protocol, reactor, error, defer, interfaces, udp
@@ -671,12 +671,16 @@
         secondPort = reactor.listenMulticast(
             portno, secondClient, listenMultiple=True)
 
-        joined = self.server.transport.joinGroup("225.0.0.250")
+        theGroup = "225.0.0.250"
+        joined = gatherResults([self.server.transport.joinGroup(theGroup),
+                                firstPort.joinGroup(theGroup),
+                                secondPort.joinGroup(theGroup)])
+
 
         def serverJoined(ignored):
             d1 = firstClient.packetReceived = Deferred()
             d2 = secondClient.packetReceived = Deferred()
-            firstClient.transport.write("hello world", ("225.0.0.250", portno))
+            firstClient.transport.write("hello world", (theGroup, portno))
             return gatherResults([d1, d2])
         joined.addCallback(serverJoined)
 
@@ -698,6 +702,7 @@
                                  "processes can listen, but not multiple sockets "
                                  "in same process?")
 
+
 if not interfaces.IReactorUDP(reactor, None):
     UDPTestCase.skip = "This reactor does not support UDP"
     ReactorShutdownInteraction.skip = "This reactor does not support UDP"


```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

It'd be nice to turn this into a ReactorBuilder test while you're at it ;)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @itamarst<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst removed owner|
|-|-|

Looks good, merge if it is green on all build slaves and you add news file.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [32857]) Branching to 'multilisten-5090'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Waiting on [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/multilisten-5090).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set status to `closed`|
|-|-|

(In [32861]) Fix multicast tests to allow for per-socket rather than per-host groups.

Author: glyph

Reviewer: itamar

Fixes: [#5090](https://github.com/twisted/twisted/issues/5090)

twisted.test.test_udp.MulticastTestCase.test_multiListen was implicitly
expecting multicast group participation to be host-wide and not per-socket.
This change causes all relevant multicast sockets being tested to join the
group, instead of relying on the host-wide behavior.  In addition to generally
being more sensible, this causes the test to no longer hang on systems with this
multicast policy.
